File: Algorithm/Python/evaluate_reverse_polish_notation.py
class Solution:
    def evalRPN(self,tokens):
        stack=[]
        for token in tokens:
           
            if token not in '+-/*':
                stack.append(int(token))
            else:
                
                if token is'+':
                    tmp=stack.pop()+stack.pop()
                    stack.append(tmp)
                elif token is '-':
                    tmp=stack[-2]-stack[-1]
                    stack=stack[:-2]
                    stack.append(tmp)
                elif token is '*':
                    tmp=stack.pop()*stack.pop()
                    stack.append(tmp)
                else:
                    if stack[-2]*stack[-1]<0:
                        tmp = -(abs(stack[-2])/abs(stack[-1]))
                    else:
                        tmp=stack[-2]/stack[-1]
                    stack=stack[:-2]
                    stack.append(tmp)
                
        top=stack.pop()
        if len(stack)==0:
            return top
        return -1
    def evalRPN1(self,tokens):
        stack=[]
        for token in tokens:
            if token not in '+-/*':
                stack.append(int(token))
            else:
                if token is '+':
                    stack.append(stack.pop() + stack.pop())
                elif token is '-':
                    stack.append(-stack.pop() + stack.pop())
                elif token is '*':
                    stack.append(stack.pop()*stack.pop())
                else:
                    tmp1 = stack.pop()
                    if tmp1 == 0:
                        return -1
                    stack.append(stack.pop()/tmp1)
        return stack[0] if len(stack) == 1 else -1

if __name__=='__main__':
    sol=Solution()
    tokens=["-3","9","*"]
    print sol.evalRPN(tokens)
    print sol.evalRPN1(tokens)
    

File: Algorithm/CPP/evaluate_reverse_polish_notation.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <string>
# include <unordered_map>
# include <cctype>
# include <algorithm>
# include <time.h>
using namespace std;



class Solution {
public:
	int evalRPN(vector<string>& tokens) {
		stack<int> Stack;
		int tmp = 0;
		int top1 = 0;
		int top2 = 0;
		for (string token : tokens)
		{
			if ((token == "+") || (token == "-") || (token == "*") || (token == "/"))
			{
				top1 = Stack.top();
				Stack.pop();
				top2 = Stack.top();
				Stack.pop();
				if (token == "+")
				{
					tmp = top2 + top1;
				}
				if (token == "-")
				{
					tmp = top2 - top1;
				}
				if (token == "*")
				{
					tmp = top2 * top1;
				}
				if (token == "/")
				{
					tmp = top2 / top1;
				}
				Stack.push(tmp);
			}
			else
			{
				Stack.push(stoi(token));
			}
		}
		return Stack.top();
	}
};
int main()
{
	Solution sol;
	vector<string> tokens = {"4","13","5","/","+"};
	cout << sol.evalRPN(tokens) << endl;
	
	system("pause");
	return 0;
}
